File: src-backend/app/models/evaluation.py
from pydantic import BaseModel


class FoldMetrics(BaseModel):
    fold: int
    r2: float
    mse: float
    mae: float


class AggregateMetrics(BaseModel):
    r2_mean: float
    r2_std: float
    mse_mean: float
    mse_std: float
    mae_mean: float
    mae_std: float


class EvaluationMetricsResponse(BaseModel):
    model_id: str
    folds: list[FoldMetrics]
    aggregate: AggregateMetrics


class PredictionPoint(BaseModel):
    actual: float
    predicted: float


class PredictionsResponse(BaseModel):
    model_id: str
    predictions: list[PredictionPoint]


class ResidualBin(BaseModel):
    range_start: float
    range_end: float
    count: int


class ResidualsResponse(BaseModel):
    model_id: str
    residuals: list[float]
    bins: list[ResidualBin]
    mean: float
    std: float


class LossCurvePoint(BaseModel):
    epoch: int
    train_loss: float
    val_loss: float


class FoldLossCurve(BaseModel):
    fold: int
    points: list[LossCurvePoint]


class LossCurveResponse(BaseModel):
    model_id: str
    folds: list[FoldLossCurve]

File: src-frontend/src/types/evaluation.ts
// src/types/evaluation.ts
export interface FoldMetrics {
  fold: number;
  r2: number;
  mse: number;
  mae: number;
}

export interface AggregateMetrics {
  r2_mean: number;
  r2_std: number;
  mse_mean: number;
  mse_std: number;
  mae_mean: number;
  mae_std: number;
}

export interface EvaluationMetricsResponse {
  model_id: string;
  folds: FoldMetrics[];
  aggregate: AggregateMetrics;
}

export interface PredictionPoint {
  actual: number;
  predicted: number;
}

export interface PredictionsResponse {
  model_id: string;
  predictions: PredictionPoint[];
}

export interface ResidualBin {
  range_start: number;
  range_end: number;
  count: number;
}

export interface ResidualsResponse {
  model_id: string;
  residuals: number[];
  bins: ResidualBin[];
  mean: number;
  std: number;
}

export interface LossCurvePoint {
  epoch: number;
  train_loss: number;
  val_loss: number;
}

export interface FoldLossCurve {
  fold: number;
  points: LossCurvePoint[];
}

export interface LossCurveResponse {
  model_id: string;
  folds: FoldLossCurve[];
}
